File: swift-bridge/Sources/MapKitBridge/LookAround.swift
import Foundation
import MapKit

struct MKRLookAroundSceneRequestStatePayload: Codable {
    var coordinate: MKRCoordinatePayload
    var hasMapItem: Bool
    var cancelled: Bool
    var loading: Bool
}

struct MKRLookAroundSnapshotOptionsPayload: Codable {
    var size: MKRScreenSizePayload
    var pointOfInterestFilter: MKRPointOfInterestFilterPayload?
}

struct MKRLookAroundSnapshotStatePayload: Codable {
    var imageByteLen: Int
    var size: MKRScreenSizePayload
}

private func mkrBuildLookAroundSnapshotOptions(
    _ payload: MKRLookAroundSnapshotOptionsPayload
) -> MKLookAroundSnapshotter.Options {
    let options = MKLookAroundSnapshotter.Options()
    options.size = mkrScreenSize(from: payload.size)
    options.pointOfInterestFilter = mkrBuildPointOfInterestFilter(payload.pointOfInterestFilter)
    return options
}

@_cdecl("mk_look_around_scene_request_new_coordinate_json")
public func mk_look_around_scene_request_new_coordinate_json(
    _ coordinateJSON: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    do {
        let coordinate = try mkrDecodeJSON(coordinateJSON, as: MKRCoordinatePayload.self)
        let request = MKLookAroundSceneRequest(coordinate: mkrCoordinate(from: coordinate))
        return mkrRetain(request)
    } catch {
        mkrSetError(outError, error)
        return nil
    }
}

@_cdecl("mk_look_around_scene_request_new_map_item_json")
public func mk_look_around_scene_request_new_map_item_json(
    _ mapItemJSON: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    do {
        let payload = try mkrDecodeJSON(mapItemJSON, as: MKRMapItemPayload.self)
        let request = MKLookAroundSceneRequest(mapItem: try mkrBuildMapItem(payload))
        return mkrRetain(request)
    } catch {
        mkrSetError(outError, error)
        return nil
    }
}

@_cdecl("mk_look_around_scene_request_state_json")
public func mk_look_around_scene_request_state_json(
    _ request: UnsafeMutableRawPointer?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    guard let request else {
        mkrSetMessageError(outError, message: "missing MKLookAroundSceneRequest")
        return nil
    }

    do {
        let sceneRequest = mkrBorrow(request, as: MKLookAroundSceneRequest.self)
        let payload = MKRLookAroundSceneRequestStatePayload(
            coordinate: mkrEncodeCoordinate(sceneRequest.coordinate),
            hasMapItem: sceneRequest.mapItem != nil,
            cancelled: sceneRequest.isCancelled,
            loading: sceneRequest.isLoading
        )
        return mkrCString(try mkrEncodeJSON(payload))
    } catch {
        mkrSetError(outError, error)
        return nil
    }
}

@_cdecl("mk_look_around_scene_request_get_scene")
public func mk_look_around_scene_request_get_scene(
    _ request: UnsafeMutableRawPointer?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    guard let request else {
        mkrSetMessageError(outError, message: "missing MKLookAroundSceneRequest")
        return nil
    }

    let sceneRequest = mkrBorrow(request, as: MKLookAroundSceneRequest.self)
    do {
        let scene: MKLookAroundScene = try mkrAwaitOnMain { completion in
            sceneRequest.getSceneWithCompletionHandler { scene, error in
                if let scene {
                    completion(.success(scene))
                } else {
                    completion(.failure(error ?? NSError(
                        domain: "mapkit-rs",
                        code: -1,
                        userInfo: [NSLocalizedDescriptionKey: "MKLookAroundSceneRequest failed without a scene"]
                    )))
                }
            }
        }
        return mkrRetain(scene)
    } catch {
        mkrSetError(outError, error)
        return nil
    }
}

@_cdecl("mk_look_around_scene_request_cancel")
public func mk_look_around_scene_request_cancel(_ request: UnsafeMutableRawPointer?) {
    guard let request else { return }
    let sceneRequest = mkrBorrow(request, as: MKLookAroundSceneRequest.self)
    sceneRequest.cancel()
}

@_cdecl("mk_look_around_scene_request_release")
public func mk_look_around_scene_request_release(_ request: UnsafeMutableRawPointer?) {
    guard let request else { return }
    mkrRelease(request)
}

@_cdecl("mk_look_around_scene_release")
public func mk_look_around_scene_release(_ scene: UnsafeMutableRawPointer?) {
    guard let scene else { return }
    mkrRelease(scene)
}

@_cdecl("mk_look_around_snapshotter_new")
public func mk_look_around_snapshotter_new(
    _ scene: UnsafeMutableRawPointer?,
    _ optionsJSON: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    guard let scene else {
        mkrSetMessageError(outError, message: "missing MKLookAroundScene")
        return nil
    }

    do {
        let lookAroundScene = mkrBorrow(scene, as: MKLookAroundScene.self)
        let payload = try mkrDecodeJSON(optionsJSON, as: MKRLookAroundSnapshotOptionsPayload.self)
        let snapshotter = MKLookAroundSnapshotter(
            scene: lookAroundScene,
            options: mkrBuildLookAroundSnapshotOptions(payload)
        )
        return mkrRetain(snapshotter)
    } catch {
        mkrSetError(outError, error)
        return nil
    }
}

@_cdecl("mk_look_around_snapshotter_get_snapshot")
public func mk_look_around_snapshotter_get_snapshot(
    _ snapshotter: UnsafeMutableRawPointer?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    guard let snapshotter else {
        mkrSetMessageError(outError, message: "missing MKLookAroundSnapshotter")
        return nil
    }

    let bridge = mkrBorrow(snapshotter, as: MKLookAroundSnapshotter.self)
    do {
        let snapshot: MKLookAroundSnapshotter.Snapshot = try mkrAwaitOnMain { completion in
            bridge.getSnapshotWithCompletionHandler { snapshot, error in
                if let snapshot {
                    completion(.success(snapshot))
                } else {
                    completion(.failure(error ?? NSError(
                        domain: "mapkit-rs",
                        code: -1,
                        userInfo: [NSLocalizedDescriptionKey: "MKLookAroundSnapshotter failed without a snapshot"]
                    )))
                }
            }
        }
        return mkrRetain(snapshot)
    } catch {
        mkrSetError(outError, error)
        return nil
    }
}

@_cdecl("mk_look_around_snapshotter_cancel")
public func mk_look_around_snapshotter_cancel(_ snapshotter: UnsafeMutableRawPointer?) {
    guard let snapshotter else { return }
    let bridge = mkrBorrow(snapshotter, as: MKLookAroundSnapshotter.self)
    bridge.cancel()
}

@_cdecl("mk_look_around_snapshotter_is_loading")
public func mk_look_around_snapshotter_is_loading(_ snapshotter: UnsafeMutableRawPointer?) -> Bool {
    guard let snapshotter else { return false }
    let bridge = mkrBorrow(snapshotter, as: MKLookAroundSnapshotter.self)
    return bridge.isLoading
}

@_cdecl("mk_look_around_snapshotter_release")
public func mk_look_around_snapshotter_release(_ snapshotter: UnsafeMutableRawPointer?) {
    guard let snapshotter else { return }
    mkrRelease(snapshotter)
}

@_cdecl("mk_look_around_snapshot_state_json")
public func mk_look_around_snapshot_state_json(
    _ snapshot: UnsafeMutableRawPointer?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    guard let snapshot else {
        mkrSetMessageError(outError, message: "missing MKLookAroundSnapshot")
        return nil
    }

    do {
        let bridge = mkrBorrow(snapshot, as: MKLookAroundSnapshotter.Snapshot.self)
        let payload = MKRLookAroundSnapshotStatePayload(
            imageByteLen: mkrImageByteLength(bridge.image),
            size: mkrEncodeScreenSize(bridge.image.size)
        )
        return mkrCString(try mkrEncodeJSON(payload))
    } catch {
        mkrSetError(outError, error)
        return nil
    }
}

@_cdecl("mk_look_around_snapshot_release")
public func mk_look_around_snapshot_release(_ snapshot: UnsafeMutableRawPointer?) {
    guard let snapshot else { return }
    mkrRelease(snapshot)
}

File: src/look_around.rs
use core::ffi::c_void;
use std::ptr::{self, NonNull};

use serde::{Deserialize, Serialize};

use crate::error::MapKitError;
use crate::ffi;
use crate::geometry::{MKCoordinate, MKScreenSize};
use crate::map_item::MKMapItem;
use crate::point_of_interest::MKPointOfInterestFilter;
use crate::private::{json_cstring, owned_handle, parse_json_ptr};

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
struct MKLookAroundSceneRequestState {
    coordinate: MKCoordinate,
    has_map_item: bool,
    cancelled: bool,
    loading: bool,
}

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
struct MKLookAroundSnapshotState {
    image_byte_len: usize,
    size: MKScreenSize,
}

/// Wraps `MKLookAroundSnapshotOptions`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MKLookAroundSnapshotOptions {
    /// Wraps `MKLookAroundSnapshotOptions.size`.
    pub size: MKScreenSize,
    /// Wraps `MKLookAroundSnapshotOptions.pointOfInterestFilter`.
    pub point_of_interest_filter: Option<MKPointOfInterestFilter>,
}

impl MKLookAroundSnapshotOptions {
    /// Creates a wrapper for `MKLookAroundSnapshotOptions`.
    pub const fn new(size: MKScreenSize) -> Self {
        Self {
            size,
            point_of_interest_filter: None,
        }
    }

    /// Wraps `MKLookAroundSnapshotOptions.pointOfInterestFilter`.
    pub fn with_point_of_interest_filter(
        mut self,
        point_of_interest_filter: MKPointOfInterestFilter,
    ) -> Self {
        self.point_of_interest_filter = Some(point_of_interest_filter);
        self
    }
}

/// Wraps `MKLookAroundSceneRequest`.
#[derive(Debug)]
pub struct MKLookAroundSceneRequest {
    raw: NonNull<c_void>,
}

impl MKLookAroundSceneRequest {
    /// Creates a wrapper for `MKLookAroundSceneRequest`.
    pub fn new(coordinate: MKCoordinate) -> Result<Self, MapKitError> {
        let coordinate = json_cstring(&coordinate, "MKCoordinate")?;
        let mut error = ptr::null_mut();
        let raw = unsafe {
            ffi::mk_look_around_scene_request_new_coordinate_json(coordinate.as_ptr(), &mut error)
        };
        let raw = owned_handle(raw, error, "failed to create MKLookAroundSceneRequest")?;
        Ok(Self { raw })
    }

    /// Wraps `MKLookAroundSceneRequest.fromMapItem`.
    pub fn from_map_item(map_item: &MKMapItem) -> Result<Self, MapKitError> {
        let map_item = json_cstring(map_item, "MKMapItem")?;
        let mut error = ptr::null_mut();
        let raw = unsafe {
            ffi::mk_look_around_scene_request_new_map_item_json(map_item.as_ptr(), &mut error)
        };
        let raw = owned_handle(
            raw,
            error,
            "failed to create MKLookAroundSceneRequest from MKMapItem",
        )?;
        Ok(Self { raw })
    }

    fn state(&self) -> Result<MKLookAroundSceneRequestState, MapKitError> {
        let mut error = ptr::null_mut();
        let payload =
            unsafe { ffi::mk_look_around_scene_request_state_json(self.raw.as_ptr(), &mut error) };
        if payload.is_null() {
            Err(unsafe {
                MapKitError::from_error_ptr(error, "failed to read MKLookAroundSceneRequest state")
            })
        } else {
            unsafe { parse_json_ptr(payload, "MKLookAroundSceneRequest state") }
        }
    }

    /// Wraps `MKLookAroundSceneRequest.coordinate`.
    pub fn coordinate(&self) -> Result<MKCoordinate, MapKitError> {
        Ok(self.state()?.coordinate)
    }

    /// Wraps `MKLookAroundSceneRequest.hasMapItem`.
    pub fn has_map_item(&self) -> Result<bool, MapKitError> {
        Ok(self.state()?.has_map_item)
    }

    /// Wraps `MKLookAroundSceneRequest.isCancelled`.
    pub fn is_cancelled(&self) -> Result<bool, MapKitError> {
        Ok(self.state()?.cancelled)
    }

    /// Wraps `MKLookAroundSceneRequest.isLoading`.
    pub fn is_loading(&self) -> Result<bool, MapKitError> {
        Ok(self.state()?.loading)
    }

    /// Wraps `MKLookAroundSceneRequest.scene`.
    pub fn scene(&self) -> Result<MKLookAroundScene, MapKitError> {
        let mut error = ptr::null_mut();
        let raw =
            unsafe { ffi::mk_look_around_scene_request_get_scene(self.raw.as_ptr(), &mut error) };
        let raw = owned_handle(raw, error, "MKLookAroundSceneRequest getScene failed")?;
        Ok(MKLookAroundScene { raw })
    }

    /// Wraps `MKLookAroundSceneRequest.cancel`.
    pub fn cancel(&self) {
        unsafe { ffi::mk_look_around_scene_request_cancel(self.raw.as_ptr()) };
    }
}

impl Drop for MKLookAroundSceneRequest {
    fn drop(&mut self) {
        unsafe { ffi::mk_look_around_scene_request_release(self.raw.as_ptr()) };
    }
}

/// Wraps `MKLookAroundScene`.
#[derive(Debug)]
pub struct MKLookAroundScene {
    raw: NonNull<c_void>,
}

impl MKLookAroundScene {
    pub(crate) const fn as_raw(&self) -> *mut c_void {
        self.raw.as_ptr()
    }
}

impl Drop for MKLookAroundScene {
    fn drop(&mut self) {
        unsafe { ffi::mk_look_around_scene_release(self.raw.as_ptr()) };
    }
}

/// Wraps `MKLookAroundSnapshotter`.
#[derive(Debug)]
pub struct MKLookAroundSnapshotter {
    raw: NonNull<c_void>,
}

impl MKLookAroundSnapshotter {
    /// Creates a wrapper for `MKLookAroundSnapshotter`.
    pub fn new(
        scene: &MKLookAroundScene,
        options: &MKLookAroundSnapshotOptions,
    ) -> Result<Self, MapKitError> {
        let options = json_cstring(options, "MKLookAroundSnapshotOptions")?;
        let mut error = ptr::null_mut();
        let raw = unsafe {
            ffi::mk_look_around_snapshotter_new(scene.as_raw(), options.as_ptr(), &mut error)
        };
        let raw = owned_handle(raw, error, "failed to create MKLookAroundSnapshotter")?;
        Ok(Self { raw })
    }

    /// Wraps `MKLookAroundSnapshotter.snapshot`.
    pub fn snapshot(&self) -> Result<MKLookAroundSnapshot, MapKitError> {
        let mut error = ptr::null_mut();
        let raw =
            unsafe { ffi::mk_look_around_snapshotter_get_snapshot(self.raw.as_ptr(), &mut error) };
        let raw = owned_handle(raw, error, "MKLookAroundSnapshotter getSnapshot failed")?;
        Ok(MKLookAroundSnapshot { raw })
    }

    /// Wraps `MKLookAroundSnapshotter.isLoading`.
    pub fn is_loading(&self) -> bool {
        unsafe { ffi::mk_look_around_snapshotter_is_loading(self.raw.as_ptr()) }
    }

    /// Wraps `MKLookAroundSnapshotter.cancel`.
    pub fn cancel(&self) {
        unsafe { ffi::mk_look_around_snapshotter_cancel(self.raw.as_ptr()) };
    }
}

impl Drop for MKLookAroundSnapshotter {
    fn drop(&mut self) {
        unsafe { ffi::mk_look_around_snapshotter_release(self.raw.as_ptr()) };
    }
}

/// Wraps `MKLookAroundSnapshot`.
#[derive(Debug)]
pub struct MKLookAroundSnapshot {
    raw: NonNull<c_void>,
}

impl MKLookAroundSnapshot {
    fn state(&self) -> Result<MKLookAroundSnapshotState, MapKitError> {
        let mut error = ptr::null_mut();
        let payload =
            unsafe { ffi::mk_look_around_snapshot_state_json(self.raw.as_ptr(), &mut error) };
        if payload.is_null() {
            Err(unsafe {
                MapKitError::from_error_ptr(error, "failed to read MKLookAroundSnapshot state")
            })
        } else {
            unsafe { parse_json_ptr(payload, "MKLookAroundSnapshot state") }
        }
    }

    /// Wraps `MKLookAroundSnapshot.imageByteLen`.
    pub fn image_byte_len(&self) -> Result<usize, MapKitError> {
        Ok(self.state()?.image_byte_len)
    }

    /// Wraps `MKLookAroundSnapshot.size`.
    pub fn size(&self) -> Result<MKScreenSize, MapKitError> {
        Ok(self.state()?.size)
    }
}

impl Drop for MKLookAroundSnapshot {
    fn drop(&mut self) {
        unsafe { ffi::mk_look_around_snapshot_release(self.raw.as_ptr()) };
    }
}
